datamanagement/miscellaneous/py2hdfread/h5_to_csv.py

```diff
@@ -0,0 +1,225 @@
+    ,'float64': 'float'
+    ,'object' : 'str'}
+
+std2pandas_types = {
+     'boolean' : 'bool'
+    ,'int'     : 'int64'
+    ,'float'   : 'float64'
+    ,'str'     : 'object'}
+
+yaml_pref = '''
+__PIPELINE_INFO__:
+  data_type: generic
+__HEADER__:
+  caller: single_cell_hmmcopy_bin
+  sample_id: SAMPLE_ID_HERE
+  file_format: csv
+'''
+
+gz = '.gz'
+csv = '.csv'
+suff_old = '.h5'
+suff_new = '%s%s' % (csv,gz)
+suff_yaml = '.yaml'
+
+class Timeit:
+    def __init__(self, name=None):
+        self.name = " '"  + name + "'" if name else ''
+    def __enter__(self):
+        self.start = time.clock()
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.t = (time.clock() - self.start) # * 1000.0
+        logger.info('Block %s; t=%f (s)' % (self.name,self.t))
+
+def diffs(x):
+    return '{}+'.format(x[0]) if x[0] == x[1] else '{}; {}'.format(x[0],x[0]-x[1])
+
+#https://stackoverflow.com/questions/17095101/outputting-difference-in-two-pandas-dataframes-side-by-side-highlighting-the-d
+def show_diff(df0,df1):
+    """
+    service function
+    """
+    df0.set_index('cell_id')
+    df1.set_index('cell_id')
+    df = pd.concat([df0, df1],axis='columns', keys=['0', '1'], join = 'outer')
+    res = df.swaplevel(axis='columns')[df0.columns[1:]]
+    x = res.groupby(level=0, axis=1).apply(lambda x: x.apply(diffs, axis=1))
+    print(x)
+
+def compare_df(df0,df1):
+    """
+    compares 2 dataframes first column names sets after this values column by column,
+    return:
+        True if df0 == df1 else False
+    """
+    cols = df0.dtypes[:] # fillna could change datatype
+
+    # panrdas imports hdf null values into a dataframe as null,
+    # after saving to csv and reading they become nan and could not be compared to null.
+    df0.fillna(value=0, inplace=True) # probably will fail if nan-s are in non numeric fields
+    df1.fillna(value=0, inplace=True)
+
+    #numeric = [float64','int64'] # in pandas bool is not subclass of int
+    #l = np.allclose(df0.select_dtypes(include=numeric), df1.select_dtypes(include=numeric))
+    #logger.info('tested: numeric; l=%i;' % (l))
+    ##compare set of column names, types will be compared later
+    l = set(df0.dtypes.keys()) == set(df1.dtypes.keys())
+    if l:
+        for name,type in cols.iteritems():
+            if 'float64' == type.name:
+                l &= np.allclose(df0[name],df1[name])
+            else:
+                l &= list(df0[name]) == list(df1[name])
+            logger.info('tested: %s; type=%s; isOk=%i;' % (name,type.name,l))
+            if not l:
+                logger.info('\n{};\n{}\n' .format(list(df0[name])[:11],list(df1[name])[:11]))
+                break
+    logger.info('etc: {};'.format(l))
+    return l
+
+
+def read_csv_with_types(filename):
+    """
+    Args:
+        filename(str): filename ends with '.csv' or '.csv.gz'
+    Returns:
+        DataFrame
+    """
+    fnm = filename[:-len(csv)] if filename.endswith(csv) else filename[:-len(suff_new)]
```

Comment:
this is a bit awkward.  assume the files are:
csv: filename
yaml: filename + '.yaml'